extensions-core/druid-kerberos/src/main/java/io/druid/kerberos/KerberosHttpClient.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.kerberos;
+
+import com.google.common.base.Strings;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.metamx.http.client.AbstractHttpClient;
+import com.metamx.http.client.HttpClient;
+import com.metamx.http.client.Request;
+import com.metamx.http.client.response.HttpResponseHandler;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.logger.Logger;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CommonConfigurationKeysPublic;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.ietf.jgss.GSSContext;
+import org.ietf.jgss.GSSException;
+import org.ietf.jgss.GSSManager;
+import org.ietf.jgss.GSSName;
+import org.ietf.jgss.Oid;
+import org.jboss.netty.handler.codec.http.HttpHeaders;
+import org.joda.time.Duration;
+
+import java.io.IOException;
+import java.security.PrivilegedExceptionAction;
+
+public class KerberosHttpClient extends AbstractHttpClient
+{
+  private static final Logger log = new Logger(KerberosHttpClient.class);
+
+  private final HttpClient delegate;
+  private final DruidKerberosConfig config;
+  private static final Base64 base64codec = new Base64(0);
+
+
+  public KerberosHttpClient(HttpClient delegate, DruidKerberosConfig config)
+  {
+    this.delegate = delegate;
+    this.config = config;
+  }
+
+  @Override
+  public <Intermediate, Final> ListenableFuture<Final> go(
+    Request request, HttpResponseHandler<Intermediate, Final> httpResponseHandler, Duration duration
+  )
+  {
+    try {
+      final String host = request.getUrl().getHost();
+      authenticateIfRequired();
+      UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
+      String challenge = currentUser.doAs(new PrivilegedExceptionAction<String>()
+      {
+        @Override
+        public String run() throws Exception
+        {
+          return kerberosChallenge(host);
```

Comment:
this should not be blocking as user is already logged in using this realm. 
Also checked how other projects implement this e.g - 
https://github.com/prongs/apache-hive/blob/master/jdbc/src/java/org/apache/hive/jdbc/HttpKerberosRequestInterceptor.java
which has similar logic underneath. 
as far as multi threading goes, Since the credentials are read from the Subject which is initiating the context. it should be safe to call this concurrently from multiple threads.